analyzer/src/main/kotlin/managers/utils/DependencyGraphBuilder.kt

```diff
@@ -0,0 +1,329 @@
+     * explicitly, as they are recorded automatically by _addDependency()_. However, if there are scopes without
+     * dependencies, this function can be used to include them into the builder result.
+     */
+    fun addScope(scopeName: String) {
+        scopeMapping.putIfAbsent(scopeName, emptyList())
+    }
+
+    /**
+     * Add the given [dependency] for the scope with the given [scopeName] to this builder. This function needs to be
+     * called all the direct dependencies of all scopes. That way the builder gets sufficient information to construct
+     * the [DependencyGraph].
+     */
+    fun addDependency(scopeName: String, dependency: D) {
+        addDependencyToGraph(scopeName, dependency, transitive = false)
+    }
+
+    /**
+     * Construct the [DependencyGraph] from the dependencies passed to this builder so far.
+     */
+    fun build(): DependencyGraph = DependencyGraph(dependencyIds, directDependencies, scopeMapping)
+
+    /**
+     * Return a set with all the packages that have been encountered for the current project.
+     */
+    fun packages(): Set<Package> = resolvedPackages
+
+    /**
+     * Update the dependency graph by adding the given [dependency], which may be [transitive], for the scope with name
+     * [scopeName]. All the dependencies of this dependency are processed recursively.
+     */
+    private fun addDependencyToGraph(scopeName: String, dependency: D, transitive: Boolean): DependencyReference {
+        val identifier = dependencyHandler.identifierFor(dependency)
+        val issues = dependencyHandler.issuesForDependency(dependency).toMutableList()
+        val index = updateDependencyMappingAndPackages(identifier, dependency, issues)
+
+        val ref = when (val result = findDependencyInGraph(index, dependency)) {
+            is DependencyGraphSearchResult.Found -> result.ref
+
+            is DependencyGraphSearchResult.NotFound -> {
+                insertIntoGraph(
+                    RootDependencyIndex(index, result.fragmentIndex),
+                    scopeName,
+                    dependency,
+                    issues,
+                    transitive
+                )
+            }
+
+            is DependencyGraphSearchResult.Incompatible ->
+                insertIntoNewFragment(index, scopeName, dependency, issues, transitive)
+        }
+
+        return updateScopeMapping(scopeName, ref, transitive)
+    }
+
+    /**
+     * Update internal state for the given dependency [id]. Check whether this ID is already known. If not, add it
+     * to the data managed by this instance, resolve the package, and update the [issues] if necessary. Return the
+     * numeric index for this dependency.
+     */
+    private fun updateDependencyMappingAndPackages(id: String, dependency: D, issues: MutableList<OrtIssue>): Int {
+        val dependencyIndex = dependencyIndexMapping[id]
+        if (dependencyIndex != null) return dependencyIndex
```

Comment:
Nit: Could be `dependencyIndexMapping[id]?.let { return it }`. But maybe that's less readable.

Reply:
I tend to be confused by returns within lambdas...